Add timer for repainting document markers.

megangardner opened the pull request:
#### 69228c8b8bcb44e81a68ed64e2fd867503d9594c
<pre>
Add timer for repainting document markers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269529">https://bugs.webkit.org/show_bug.cgi?id=269529</a>
<a href="https://rdar.apple.com/123048192">rdar://123048192</a>

Reviewed by Aditya Keerthi.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::DocumentMarkerController):
(WebCore::DocumentMarkerController::detach):
(WebCore::DocumentMarkerController::addMarker):
(WebCore::DocumentMarkerController::removeMarkers):
(WebCore::DocumentMarkerController::unifiedTextReplacementAnimationTimerFired):
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::drawUnifiedTextReplacementUnderline):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarker):

Canonical link: <a href="https://commits.webkit.org/274914@main">https://commits.webkit.org/274914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/869950c73eb8579b7d595d0291b7463c165e42bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40285 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19297 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36374 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42592 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22227 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16626 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33451 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40859 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16250 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34764 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14026 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14072 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35717 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36524 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36047 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39788 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15100 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12371 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38065 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16719 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9057 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16768 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16363 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -45,6 +45,8 @@ namespace WebCore {
 constexpr Seconds markerFadeAnimationDuration = 200_ms;
 constexpr double markerFadeAnimationFrameRate = 30;
 
+constexpr double unifiedTextReplacementAnimationFrameRate = 30;
```
Can we merge with `markerFadeAnimationFrameRate`?

pxlcoder commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -1034,14 +1035,30 @@ FloatRect LegacyTextBoxPainter::calculateUnionOfAllDocumentMarkerBounds(const Le
+
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    if (markedText.type == MarkedText::Type::UnifiedTextReplacement)
+        drawUnifiedTextReplacementUnderline(m_paintInfo.context(), bounds,  m_renderer.frame().view()->size());
```
Should there be an early return here?

megangardner replied:
yes, you're right, thanks for catching.

aprotyas commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -711,6 +726,17 @@ void DocumentMarkerController::dismissMarkers(OptionSet<DocumentMarker::Type> ty
+void DocumentMarkerController::unifiedTextReplacementAnimationTimerFired()
+{
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    forEachOfTypes({ DocumentMarker::Type::UnifiedTextReplacement }, [&] (Node& node, RenderedDocumentMarker&) {
```
It looks like we're not using any captured value, so we can just drop `&`? Also small nit, style guide says no space between capture list and arg list.
```suggestion
    forEachOfTypes({ DocumentMarker::Type::UnifiedTextReplacement }, [](Node& node, RenderedDocumentMarker&) {
```

aprotyas reviewed (commented):
LGTM, one comment

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -380,6 +391,10 @@ void DocumentMarkerController::copyMarkers(Node& source, OffsetRange range, Node
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    if (types.contains(DocumentMarker::Type::UnifiedTextReplacement))
+        m_unifiedTextReplacementAnimationTimer.stop();
+#endif
```
I don't think this is right, because this method could only remove the markers on a single node, and not across the entire document.

`DocumentMarkerController::removeMarkers(OptionSet<DocumentMarker::Type> types, const Function<FilterMarkerResult(const RenderedDocumentMarker&)>& filter)` is probably a better place for this.

pxlcoder commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -560,6 +570,11 @@ void DocumentMarkerController::removeMarkers(OptionSet<DocumentMarker::Type> typ
+#if ENABLE(UNIFIED_TEXT_REPLACEMENT)
+    if (types.contains(DocumentMarker::Type::UnifiedTextReplacement))
+        m_unifiedTextReplacementAnimationTimer.stop();
+#endif
```
This should be moved after the loop and consult `removedMarkerTypes` rather than `types`. The filter function might result in nothing being removed.

megangardner committed:
Add timer for repainting document markers.
https://bugs.webkit.org/show_bug.cgi?id=269529
rdar://123048192

Reviewed by Aditya Keerthi.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::DocumentMarkerController):
(WebCore::DocumentMarkerController::detach):
(WebCore::DocumentMarkerController::addMarker):
(WebCore::DocumentMarkerController::removeMarkers):
(WebCore::DocumentMarkerController::unifiedTextReplacementAnimationTimerFired):
* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/rendering/MarkedText.cpp:
(WebCore::MarkedText::collectForDocumentMarkers):
* Source/WebCore/rendering/MarkedText.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::drawUnifiedTextReplacementUnderline):
(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarker):

Canonical link: https://commits.webkit.org/274914@main
